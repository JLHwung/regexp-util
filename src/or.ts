import { Base } from './base';
import { Charset, CharsetInput } from './charset';

export type OrInput = string | CharsetInput;

export class Or extends Base {
  public charset: Charset;
  public strings: Set<string>;

  constructor(...inputs: OrInput[]) {
    super();

    this.strings = new Set();
    const charset_inputs: CharsetInput[] = [];

    for (const input of inputs) {
      if (typeof input !== 'string' || input.length === 1) {
        charset_inputs.push(input);
      } else {
        this.strings.add(input);
      }
    }

    this.charset = new Charset(...charset_inputs);
  }

  public union(...inputs: OrInput[]) {
    return new Or(this.charset, ...this.strings, ...inputs);
  }

  protected _is_empty() {
    return this.charset.isEmpty() && this.strings.size === 0;
  }

  protected _to_string() {
    const parts: string[] = [];

    if (this.charset.data.length !== 0) {
      parts.push(this.charset.toString());
    }

    if (this.strings.size !== 0) {
      parts.push(...this.strings);
    }

    return `(?:${parts.join('|')})`;
  }
}

export const or = (...inputs: OrInput[]) => new Or(...inputs);
