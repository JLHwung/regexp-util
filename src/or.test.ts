import { or } from './or';

test('create: accept string', () => {
  expect(or('hello', 'world').toString()).toMatchSnapshot();
});

test('create: accept charset inputs', () => {
  expect(or('a', [1, 3]).toString()).toMatchSnapshot();
});

test('create: accept or', () => {
  expect(or('a', [1, 3], or('c', 'd')).toString()).toMatchSnapshot();
});

test('union', () => {
  expect(
    or('hello', 'world', [5, 9])
      .union([3, 7])
      .toString(),
  ).toMatchSnapshot();
});

test('subtract', () => {
  expect(
    or('hello', 'world', [1, 7], 'foo', 'bar')
      .subtract('world', [3, 5], or('bar', 'baz'))
      .toString(),
  ).toMatchSnapshot();
});
