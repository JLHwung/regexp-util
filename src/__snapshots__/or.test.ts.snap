// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create: accept charset inputs 1`] = `"(?:[\\\\u{1}-\\\\u{3}\\\\u{61}])"`;

exports[`create: accept or 1`] = `"(?:[\\\\u{1}-\\\\u{3}\\\\u{61}\\\\u{63}-\\\\u{64}])"`;

exports[`create: accept string 1`] = `"(?:hello|world)"`;

exports[`subtract 1`] = `"(?:[\\\\u{1}-\\\\u{2}\\\\u{6}-\\\\u{7}]|hello|foo)"`;

exports[`union 1`] = `"(?:[\\\\u{3}-\\\\u{9}]|hello|world)"`;
